Treat Kotlin nullable as non-required

rocketraman opened the pull request:
Modify the core MethodParameter class to return the Kotlin nullability
of a parameteter (cached for performance).

Where `isOptional` is used, also check for `isKotlinNullable` i.e.
values are not considered required if they are Kotlin nullables:
- spring-messaging: named value method arguments
- spring-web: named value method arguments
- spring-webmvc: request parts

This means that Kotlin client code no longer has to explicity specify
"required=false" for Kotlin nullables -- this information is inferred
automatically by the framework.

Issue: SPR-14165

rocketraman commented:
Notes on submission / pending work:

* I put the kotlin-specific code into a separate module `spring-core-kotlin` to avoid adding the kotlin compiler to the `spring-core` module (inspired from `spring-beans-groovy`).
* I wasn't sure where to add tests for this. My first though was `MethodParameterTests`, but currently it contains very little in the way of tests. Guidance is appreciated.

Otherwise the implementation turned out to be quite simple and concise.

sdeleuze commented:
Thanks for this PR, I only had a quick look but I would be in favor of not creating a new `spring-core-kotlin`, instead you could add kotlin dependency as an optional one. After that I will make a detailed review of your PR.

rocketraman commented:
> Thanks for this PR, I only had a quick look but I would be in favor of not creating a new spring-core-kotlin, instead you could add kotlin dependency as an optional one. After that I will make a detailed review of your PR.

Done!

rocketraman commented:
NOTE: Another place we could use this functionality is when `@Inject` or `@Autowired` is used, the Kotlin nullable information could be used to determine if the value is required or not.

rocketraman commented:
@sdeleuze Ping. Looks like it needs another rebase, but would appreciate a review.

sdeleuze commented on spring-core/src/main/java/org/springframework/core/MethodParameter.java:
```diff
@@ -329,6 +338,19 @@ public MethodParameter nestedIfOptional() {
+	/**
+	 * Return whether this method parameter is declared as a Kotlin
+	 * nullable.
+	 * @since 4.4
```
Let's target 5.0 for this feature

sdeleuze commented on spring-core/src/main/java/org/springframework/core/MethodParameter.java:
```diff
@@ -329,6 +338,19 @@ public MethodParameter nestedIfOptional() {
+	 * nullable.
+	 * @since 4.4
+	 */
+	public boolean isKotlinNullable() {
```
Maybe `isNullable()`would be less specific and let the door opened to supporting this is other languages or with specific annotations

sdeleuze commented on spring-core/src/main/java/org/springframework/core/MethodParameter.java:
```diff
@@ -50,6 +51,12 @@
  */
 public class MethodParameter {
 
+	private static final boolean kotlinPresent;
```
Let's encapsulate that in a KotlinUtils class

sdeleuze commented on spring-core/src/main/java/org/springframework/core/MethodParameter.java:
```diff
@@ -67,6 +74,8 @@
 
 	private volatile Class<?> parameterType;
 
+	private volatile Boolean isKotlinNullable;
```
I think I would try initially without caching the result, unless you have specific concern about the performances

sdeleuze commented on spring-core/src/main/kotlin/org/springframework/core/MethodParameterKotlin.kt:
```diff
@@ -0,0 +1,49 @@
+ * @since 4.4
+ * @see MethodParameter
+ */
+object MethodParameterKotlin {
```
Could we turn this into a regular java `KotlinUtils` class with maybe an additional method to know if Kotlin is present? I think I would like to restrict Kotlin code to only test classes.

sdeleuze commented:
Sorry for the delay and thanks for your updated PR. I have added comments on the relevant lines of your commit.

Could you also rebase that on top of the new master (Spring Framework 5.0) and add some Kotlin unit tests to check that works?

Good idea to support that also for `@Inject`, feel free to add that to your PR and update the title of this JIRA issue to make is less specific.

rocketraman commented:
> Good idea to support that also for @Inject, feel free to add that to your PR and update the title of this JIRA issue to make is less specific.

It looks more complicated to do this [1], so I'd prefer to have this pull request reviewed as-is for now, just to keep things as simple as possible. We can add the `@Inject` functionality in a subsequent pull.

> Sorry for the delay and thanks for your updated PR. I have added comments on the relevant lines of your commit.
> 
> Could you also rebase that on top of the new master (Spring Framework 5.0) and add some Kotlin unit tests to check that works?

I have updated the pull with all of your comments taken into account. Thanks!

[1] It looks like a variation of `RequiredAnnotationBeanPostProcessor` is needed.

rocketraman committed:
Treat Kotlin nullable as non-required

Where `isOptional` is used, also check for `isKotlinNullable` i.e.
values are not considered required if they are Kotlin nullables:

- spring-messaging: named value method arguments
- spring-web: named value method arguments
- spring-webmvc: request parts

This means that Kotlin client code no longer has to explicity specify
"required=false" for Kotlin nullables -- this information is inferred
automatically by the framework.

At this time, this introspection does not apply to bean injection --
only to web and messaging method arguments.

Issue: SPR-14165

rocketraman commented:
Ping @sdeleuze for review.

rocketraman committed:
Kotlin utils and MethodParameter nullability

Modify the core MethodParameter class to return the Kotlin nullability
of a parameteter (cached for performance). The Kotlin-specific
reflection code is within a new KotlinUtils object. Calls
to this code are gated by the presence of Kotlin in the classpath,
which is checked once at initialization time.

Issue: SPR-14165

sdeleuze commented:
Superseeded by #1242
